Mac OS X build error (template/types)

Hi there,

I have been able to build Adiak on Mac OS X fine, but when I try to use it as third party library in another code, I am getting a template type conversion error. Here is the console output for the build:

`In file included from /Users/j0529096/Documents/code/thirdPartyLibs/for_caliper_install/adiak/include/adiak.hpp:52:
/Users/j0529096/Documents/code/thirdPartyLibs/for_caliper_install/adiak/include/adiak_internal.hpp:153:40: error: cannot
      initialize a parameter of type 'void *' with an lvalue of type 'long long'
            element_type<T>::set(*val, obj);
                                       ^~~
/Users/j0529096/Documents/code/thirdPartyLibs/for_caliper_install/adiak/include/adiak.hpp:129:48: note: in instantiation of
      member function 'adiak::internal::parse<long long>::make_value' requested here
      bool result = adiak::internal::parse<T>::make_value(value, avalue, datatype);
                                               ^
/Users/j0529096/Documents/code/GEOSX/src/coreComponents/managers/initialization.hpp:155:10: note: in instantiation of
      function template specialization 'adiak::value<long long>' requested here
  adiak::value( name + " sum", total );
         ^
/Users/j0529096/Documents/code/GEOSX/src/coreComponents/managers/initialization.cpp:184:3: note: in instantiation of
      function template specialization 'geosx::pushStatsIntoAdiak<long long>' requested here
  pushStatsIntoAdiak( "numThreads", numThreads );
  ^
/Users/j0529096/Documents/code/thirdPartyLibs/for_caliper_install/adiak/include/adiak_internal.hpp:37:50: note: passing
      argument to parameter 'p' here
         static void set(adiak_value_t &v, void *p) { v.v_ptr = p; }
                                                 ^
1 error generated.`

This is a Release build, with AppleClang 11.0.3.11030032 and Adiak-0.2.0.

Thanks for your help on this,

Herve

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -31,6 +31,9 @@ endif()
 include(${BLT_SOURCE_DIR}/SetupBLT.cmake)
 
 add_subdirectory(src)
+if (ENABLE_TESTS)
+  add_subdirectory(tests)
+endif()
 
 include(CMakePackageConfigHelpers)
 

--- a/include/adiak.h
+++ b/include/adiak.h
@@ -13,8 +13,8 @@ extern "C" {
 #endif
 
 #define ADIAK_VERSION 0
-#define ADIAK_MINOR_VERSION 2
-#define ADIAK_POINT_VERSION 1
+#define ADIAK_MINOR_VERSION 3
+#define ADIAK_POINT_VERSION 0
 
 // ADIAK DATA CATEGORIES 
 // Please treat values through 1000 as reserved. 1001 onward can be used as
@@ -35,7 +35,7 @@ typedef enum {
 
 typedef enum {
    adiak_type_unset = 0,
-   adiak_long,        /* adiak_value string: %ld (implies adiak_rational) */
+   adiak_long,        /* adiak_value string: %ld (implies rational) */
    adiak_ulong,       /* adiak_value string: %lu (implies rational) */   
    adiak_int,         /* adiak_value string: %d (implies rational) */
    adiak_uint,        /* adiak_value string: %u (implies rational) */
@@ -49,7 +49,9 @@ typedef enum {
    adiak_range,       /* adiak_value string <typestring> (implies categorical), */
    adiak_set,         /* adiak_value string [typestring], passed as a size integer */
    adiak_list,        /* adiak_value string {typestring}, passed as a size integer */
-   adiak_tuple        /* adiak_value string (typestring1, typestring2, ..., typestringN), passed as an N integer */
+   adiak_tuple,       /* adiak_value string (typestring1, typestring2, ..., typestringN), passed as an N integer */
+   adiak_longlong,    /* adiak_value string: %lld (implies rational) */
+   adiak_ulonglong    /* adiak_value string: %llu (implies rational) */
 } adiak_type_t;
 
 typedef int adiak_category_t;
@@ -70,6 +72,7 @@ typedef union adiak_value_t {
    double v_double;
    void *v_ptr;
    union adiak_value_t *v_subval;
+   long long v_longlong;
 } adiak_value_t;
    
 static const adiak_datatype_t adiak_unset_datatype = { adiak_type_unset, adiak_numerical_unset, 0, 0, NULL };

--- a/include/adiak_internal.hpp
+++ b/include/adiak_internal.hpp
@@ -44,6 +44,14 @@ namespace adiak
          static const adiak_type_t dtype = adiak_ulong;
          static void set(adiak_value_t &v, unsigned long p) { v.v_long = p; }
       };
+      template<> struct element_type<long long> {
+         static const adiak_type_t dtype = adiak_longlong;
+         static void set(adiak_value_t &v, long long p) { v.v_longlong = p; }
+      };
+      template<> struct element_type<unsigned long long> {
+         static const adiak_type_t dtype = adiak_ulonglong;
+         static void set(adiak_value_t &v, unsigned long long p) { v.v_longlong = p; }
+      };
       template<> struct element_type<int> {
          static const adiak_type_t dtype = adiak_int;
          static void set(adiak_value_t &v, int p) { v.v_int = p; }

--- a/src/adiak.c
+++ b/src/adiak.c
@@ -55,6 +55,8 @@ static int measure_adiak_cputime;
 
 static adiak_datatype_t base_long = { adiak_long, adiak_rational, 0, 0, NULL };
 static adiak_datatype_t base_ulong = { adiak_ulong, adiak_rational, 0, 0, NULL };
+static adiak_datatype_t base_longlong = { adiak_longlong, adiak_rational, 0, 0, NULL };
+static adiak_datatype_t base_ulonglong = { adiak_ulonglong, adiak_rational, 0, 0, NULL };
 static adiak_datatype_t base_int = { adiak_int, adiak_rational, 0, 0, NULL };
 static adiak_datatype_t base_uint = { adiak_uint, adiak_rational, 0, 0, NULL };
 static adiak_datatype_t base_double = { adiak_double, adiak_rational, 0, 0, NULL };
@@ -152,6 +154,10 @@ int adiak_namevalue(const char *name, int category, const char *subcategory, con
       case adiak_uint:
          value->v_int = va_arg(ap, int);
          break;
+      case adiak_longlong:
+      case adiak_ulonglong:
+         value->v_longlong = va_arg(ap, long long);
+         break;
       case adiak_double:
          value->v_double = va_arg(ap, double);
          break;
@@ -196,6 +202,8 @@ adiak_numerical_t adiak_numerical_from_type(adiak_type_t dtype)
          return adiak_numerical_unset;
       case adiak_long:
       case adiak_ulong:
+      case adiak_longlong:
+      case adiak_ulonglong:
       case adiak_int:
       case adiak_uint:
       case adiak_double:
@@ -371,6 +379,11 @@ static adiak_type_t toplevel_type(const char *typestr) {
             cur++;
             if (*cur == 'd') return adiak_long;
             if (*cur == 'u') return adiak_ulong;
+            if (*cur == 'l') {
+               cur++;
+               if (*cur == 'd') return adiak_longlong;
+               if (*cur == 'u') return adiak_ulonglong;
+            }
             return adiak_type_unset;
          case 'd': return adiak_int;
          case 'u': return adiak_uint;
@@ -403,6 +416,10 @@ adiak_datatype_t *adiak_get_basetype(adiak_type_t t)
          return &base_long;
       case adiak_ulong:
          return &base_ulong;
+      case adiak_longlong:
+         return &base_longlong;
+      case adiak_ulonglong:
+         return &base_ulonglong;
       case adiak_int:
          return &base_int;
       case adiak_uint:
@@ -453,6 +470,8 @@ static void free_adiak_value_worker(adiak_datatype_t *t, adiak_value_t *v) {
       case adiak_type_unset:
       case adiak_long:
       case adiak_ulong:
+      case adiak_longlong:
+      case adiak_ulonglong:
       case adiak_int:
       case adiak_uint:
       case adiak_double:
@@ -500,6 +519,10 @@ static int copy_value(adiak_value_t *target, adiak_datatype_t *datatype, void *p
       case adiak_date:
          target->v_long = *((long *) ptr);
          return sizeof(long);
+      case adiak_longlong:
+      case adiak_ulonglong:
+         target->v_longlong = *((long long *) ptr);
+         return sizeof(long long);
       case adiak_int:
       case adiak_uint:
          target->v_int = *((int *) ptr);
@@ -545,7 +568,7 @@ static adiak_datatype_t *parse_typestr_helper(const char *typestr, int typestr_s
 {
    adiak_datatype_t *t = NULL;
    int cur = typestr_start;
-   int end_brace, i, is_long = 0;
+   int end_brace, i, is_long = 0, is_longlong = 0;
    
    if (!typestr)
       goto error;
@@ -611,13 +634,17 @@ static adiak_datatype_t *parse_typestr_helper(const char *typestr, int typestr_s
       if (typestr[cur] == 'l') {
          is_long = 1;
          cur++;
+         if (typestr[cur] == 'l') {
+            is_longlong = 1;
+            cur++;
+         }
       }
       switch (typestr[cur]) {
          case 'd':
-            t = is_long ? &base_long : &base_int;
+            t = is_long ? (is_longlong ? &base_longlong  : &base_long)  : &base_int;
             break;
          case 'u':
-            t = is_long ? &base_ulong : &base_uint;
+            t = is_long ? (is_longlong ? &base_ulonglong : &base_ulong) : &base_uint;
             break;
          case 'f':
             t = &base_double;
@@ -1074,6 +1101,12 @@ static int adiak_type_string_helper(adiak_datatype_t *t, char *str, int len, int
          break;
       case adiak_ulong:
          simple = long_form ? "unsigned long" : "%lu";
+         break;
+      case adiak_longlong:
+         simple = long_form ? "long long" : "%lld";
+         break;
+      case adiak_ulonglong:
+         simple = long_form ? "unsigned long long" : "%llu";
          break;
       case adiak_int:
          simple = long_form ? "int" : "%d";

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,0 +1,14 @@
+include_directories(.)
+include_directories(../include)
+
+blt_add_library( NAME testlib
+                 SOURCES testlib.c
+                 DEFINES "TOOLNAME=TOOL3"
+                 DEPENDS_ON adiak)
+
+blt_add_executable( NAME testapp
+                    SOURCES testapp.c
+                    DEPENDS_ON testlib adiak dl)
+blt_add_executable( NAME testapp-cxx
+                    SOURCES testapp.cpp
+                    DEPENDS_ON testlib adiak dl)

--- a/tests/testapp.c
+++ b/tests/testapp.c
@@ -25,6 +25,9 @@ void dowork(struct timeval start)
    result = adiak_namevalue("problemsize", adiak_general, NULL, "%lu", 14000);
    if (result != 0) printf("return: %d\n\n", result);
 
+   result = adiak_namevalue("countdown", adiak_general, NULL, "%lld", 9876543210);
+   if (result != 0) printf("return: %d\n\n", result);
+
    result = adiak_user();
    if (result != 0) printf("return: %d\n\n", result);
    

--- a/tests/testapp.cpp
+++ b/tests/testapp.cpp
@@ -73,6 +73,9 @@ void dowork(struct timeval start)
    result = adiak::value("problemsize", 14000);
    if (!result) printf("return: %d\n\n", result);
 
+   result = adiak::value("countdown", static_cast<long long>(9876543210));
+   if (!result) printf("return: %d\n\n", result);
+
    result = adiak::value("allnames", names);
    if (!result) printf("return: %d\n\n", result);
 

--- a/tests/testlib.c
+++ b/tests/testlib.c
@@ -23,6 +23,12 @@ static void print_value(adiak_value_t *val, adiak_datatype_t *t)
       case adiak_ulong:
          printf("%lu", (unsigned long) val->v_long);
          break;
+      case adiak_longlong:
+         printf("%lld", val->v_longlong);
+         break;
+      case adiak_ulonglong:
+         printf("%llu", (unsigned long long) val->v_longlong);
+         break;
       case adiak_int:
          printf("%d", val->v_int);
          break;

